src/docfx/lib/ParallelUtility.cs

```diff
@@ -54,29 +55,29 @@ public static Task ForEach<T>(IEnumerable<T> source, Func<T, Task> action, Actio
         /// </summary>
         public static Task ForEach<T>(IEnumerable<T> source, Func<T, Action<T>, Task> action, Func<T, bool> predicate = null, Action<int, int> progress = null)
         {
-            ActionBlock<T> queue = null;
+            ActionBlock<(T item, IEnumerable<T> inputs)> queue = null;
 
-            var total = 0;
             var done = 0;
-            var running = 0;
-
-            queue = new ActionBlock<T>(Run, s_dataflowOptions);
-
-            foreach (var item in source)
-            {
-                Enqueue(item);
-            }
-
-            if (Volatile.Read(ref running) == 0)
-            {
-                queue.Complete();
-            }
+            var total = 0;
+            var running = 1;
 
+            queue = new ActionBlock<(T, IEnumerable<T>)>(Run, s_dataflowOptions);
+            queue.Post((default, source));
             return queue.Completion;
 
-            async Task Run(T item)
+            async Task Run((T item, IEnumerable<T> inputs) item)
             {
-                await action(item, Enqueue);
+                if (item.inputs != null)
```

Comment:
>item [](start = 20, length = 4)

`item?.inputs`?

Reply:
or `inputs`?

---
In reply to: [199336848](https://github.com/dotnet/docfx/pull/2976#discussion_r199336848) [](ancestors = 199336848)